src/amd/compiler/aco_register_allocation.cpp

```diff
@@ -1005,13 +1007,13 @@ void register_allocation(Program *program, std::vector<std::set<Temp>> live_out_
       assert(info->second.block_idx != 0);
       Instruction* instr = info->second.phi;
       Temp same = Temp();
-      Definition def = instr->getDefinition(0);
+      Definition def = instr->definitions[0];
```

Comment:
what does .clear() actually do? Because I get a segfault in one shader this way and when running with assertions:

```
run: ../../../../../../Projects/mesa/src/amd/compiler/aco_util.h:153: aco::span<T>::value_type& aco::span<T>::operator[](aco::span<T>::size_type) [with T = aco::Definition; aco::span<T>::reference = aco::Definition&; aco::span<T>::value_type = aco::Definition; aco::span<T>::size_type = long unsigned int]: Assertion `length > index' failed.
==20471== 
==20471== Process terminating with default action of signal 6 (SIGABRT)
==20471==    at 0x4906ED7: raise (raise.c:51)
==20471==    by 0x48E8534: abort (abort.c:79)
==20471==    by 0x48E840E: __assert_fail_base.cold.0 (assert.c:92)
==20471==    by 0x48F8011: __assert_fail (assert.c:101)
==20471==    by 0x5142672: aco::span<aco::Definition>::operator[](unsigned long) [clone .isra.227] [clone .part.228] (aco_util.h:153)
==20471==    by 0x5144F6E: operator[] (aco_ir.h:320)
==20471==    by 0x5144F6E: operator() (aco_register_allocation.cpp:1013)
==20471==    by 0x5144F6E: std::_Function_handler<aco::Temp (std::_Rb_tree_iterator<std::pair<unsigned int const, aco::register_allocation(aco::Program*, std::vector<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> >, std::allocator<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> > > >)::phi_info> >), aco::register_allocation(aco::Program*, std::vector<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> >, std::allocator<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> > > >)::{lambda(std::_Rb_tree_iterator<std::pair<unsigned int const, aco::register_allocation(aco::Program*, std::vector<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> >, std::allocator<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> > > >)::phi_info> >)#3}>::_M_invoke(std::_Any_data const&, std::_Rb_tree_iterator<std::pair<unsigned int const, aco::register_allocation(aco::Program*, std::vector<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> >, std::allocator<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> > > >)::phi_info> >&&) (std_function.h:282)
==20471==    by 0x514C13F: operator() (std_function.h:687)
==20471==    by 0x514C13F: aco::register_allocation(aco::Program*, std::vector<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> >, std::allocator<std::set<aco::Temp, std::less<aco::Temp>, std::allocator<aco::Temp> > > >) (aco_register_allocation.cpp:1822)
==20471==    by 0x5132514: aco_compile_shader (aco_interface.cpp:74)
==20471==    by 0x511CEFB: shader_variant_compile (radv_shader.c:1222)
==20471==    by 0x511DB9D: radv_shader_variant_compile (radv_shader.c:1285)
==20471==    by 0x510C2C1: radv_create_shaders (radv_pipeline.c:2553)
==20471==    by 0x51111F4: radv_pipeline_init (radv_pipeline.c:4458)
```
I'm not completely sure how this happens, though. If I assert(instr->num_definitions) on master at this position, it doesn't hit it. (see l. 1050 how we indicate trivial phis, although that is not really a good solution and reminds me to rework SSA-reconstruction).

Reply:
I also checked definitions.size() and it is 1. I can also easily print the instruction right before and it has a definition. Does that make any sense?

Reply:
Ah the problem is that clear resets `data` to ``nullptr`.

It seems that the instruction is indeed reused and tries to access the now invalid pointer. The fix would be to simply remove that assignment and only set length to 0

Reply:
no, I actually think that is okay as it uncovers bugs :P
Seems that valgrind messed up the line numbers and the segfault happened further down (didn't know that valgrind is also fragile with equally named variables). Anyway, I'll push a fix to backend to avoid accessing definitions[0] if num_definitions was already set to zero.

Reply:
done. can you rebase one last time? and hopefully, then we can land this.